libimage/push_test.go

```diff
@@ -96,3 +100,97 @@ func TestPushOtherPlatform(t *testing.T) {
 	_, err = runtime.Push(ctx, "docker.io/library/alpine:latest", "docker-archive:"+tmp.Name(), pushOptions)
 	require.NoError(t, err)
 }
+
+func TestPushWithForceCompression(t *testing.T) {
+	runtime, cleanup := testNewRuntime(t)
+	defer cleanup()
+	ctx := context.Background()
+
+	// Prefetch alpine.
+	pullOptions := &PullOptions{}
+	pullOptions.Writer = os.Stdout
+	pullOptions.Architecture = "arm64"
+	pulledImages, err := runtime.Pull(ctx, "docker.io/library/alpine:latest", config.PullPolicyAlways, pullOptions)
+	require.NoError(t, err)
+	require.Len(t, pulledImages, 1)
+
+	data, err := pulledImages[0].Inspect(ctx, nil)
+	require.NoError(t, err)
+	require.Equal(t, "arm64", data.Architecture)
+
+	// Push newly pulled alpine to directory with uncompressed blobs
+	pushOptions := &PushOptions{}
+	pushOptions.SystemContext = &types.SystemContext{}
+	pushOptions.SystemContext.DirForceDecompress = true
+	pushOptions.Writer = os.Stdout
+	dirDest := t.TempDir()
+	_, err = runtime.Push(ctx, "docker.io/library/alpine:latest", "dir:"+dirDest, pushOptions)
+	require.NoError(t, err)
+
+	// Pull uncompressed alpine from `dir:dirDest` as source.
+	pullOptions = &PullOptions{}
+	pullOptions.Writer = os.Stdout
+	pullOptions.Architecture = "arm64"
+	pulledImages, err = runtime.Pull(ctx, "dir:"+dirDest, config.PullPolicyAlways, pullOptions)
+	require.NoError(t, err)
+	require.Len(t, pulledImages, 1)
```

Comment:
FWIW I didn’t test this now but I don’t think the `dir:` step is necessary if the test is structured primarily around c/storage.

c/storage itself is a source of always-uncompressed data.

(This works well enough as is, no need to go back and spend time changing it.)